Tools/src/ValidationTool/ValidationTool.cs

```diff
@@ -100,125 +113,227 @@ private static string[] ListSpecFiles(string specsPath)
             return paths;
         }
 
-        private static int RunPipeline(string filePath)
+        private static void SaveMatchResults(List<BasicStageResult> stageResults, string path)
         {
-            Console.WriteLine(SplitLineStr);
-            Console.WriteLine($"Begin to check file: {filePath}");
+            if (stageResults.Count <= 1)
+            {
+                return;
+            }
+
+            string newPath = path + ".bak";
+            File.Copy(path, newPath);
+
+            // Convert match results and original specs to same data type Newtonsoft.Json.Linq.JArray.
+            var rawSpecs = JsonConvert.SerializeObject(stageResults[stageResults.Count - 1].Specs);
+            List<dynamic> jsonSpecs = (List<dynamic>)JsonConvert.DeserializeObject<IList<dynamic>>(rawSpecs);
+            var rawData = File.ReadAllText(path);
+            List<dynamic> originalSpecs = (List<dynamic>)JsonConvert.DeserializeObject<IList<dynamic>>(rawData);
+
+            // Replace matching results.
+            for (int i = 0; i < jsonSpecs.Count; ++i)
+            {
+                originalSpecs[i]["Results"] = jsonSpecs[i]["Results"];
+            }
+
+            string rawResults = JsonConvert.SerializeObject(originalSpecs, Formatting.Indented);
+            File.WriteAllText(path, rawResults);
```

Comment:
We should specify UTF-8 as the encoding here. The project assumes all specs are utf-8.